sdk/keyvault/keyvault-keys/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -82,94 +101,82 @@ export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
-        const { authorization, resource } = this.parseWWWAuthenticate(www_authenticate);
-        const challenge = new AuthenticationChallenge(authorization, resource + "/.default")
-
-        if (!challenge.equalTo(this.challengeCache.challenge)) {
-          this.challengeCache.setCachedChallenge(challenge);
-          this.tokenCache.setCachedToken(undefined);
-
-          await this.authenticateRequest(webResource);
-          return this._nextPolicy.sendRequest(webResource);
-        }
-        return response;
-      }
+    // If we don't receive a response with a 401 status code,
+    // then we can assume this response has nothing to do with the challenge authentication process.
+    if (response.status !== 401) {
       return response;
-    } else {
+    }
+
+    // If the response status is 401, we only re-authenticate if the WWW-Authenticate header is present.
+    const wwwAuthenticate = response.headers.get("WWW-Authenticate");
+    if (!wwwAuthenticate) {
       return response;
     }
+
+    // The challenge based authentication will contain both:
+    // - An authorization URI with a token,
+    // - The resource to which that token is valid against (also called the scope).
+    const parsedWWWAuth = this.parseWWWAuthenticate(wwwAuthenticate);
+    const authorization = parsedWWWAuth.authorization! || parsedWWWAuth.authorization_url!;
+    const resource = parsedWWWAuth.resource! || parsedWWWAuth.scope!;
+
+    const challenge = new AuthenticationChallenge(authorization, resource + "/.default");
+
+    // Either if there's no cached challenge at this point (could have happen in parallel),
+    // or if the cached challenge has a different scope,
+    // we store the just received challenge and reset the cached token, to force a re-authentication.
+    // This is exactly what C# is doing, as we can see here:
+    // https://github.com/heaths/azure-sdk-for-net/blob/a35434c1ce955d4cdeb393748f9f6407d4a984e2/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L233
+    if (!this.challengeCache.challenge?.equalTo(challenge)) {
+      this.challengeCache.setCachedChallenge(challenge);
+      this.tokenCache.setCachedToken(undefined);
+    }
+
+    await this.loadToken(webResource);
+    return this._nextPolicy.sendRequest(webResource);
   }
 
-  private async authenticateRequest(webResource: WebResource): Promise<void> {
+  /**
+   * Gets or updates the token from the token cache into the headers of the received web resource.
+   */
+  private async loadToken(webResource: WebResource): Promise<void> {
     let accessToken = this.tokenCache.getCachedToken();
+
+    // If there's no cached token in the cache, we try to get a new one.
     if (accessToken === undefined) {
-      accessToken = (await this.credential.getToken(this.challengeCache.challenge!.scope)) || undefined;
+      const receivedToken = await this.credential.getToken(this.challengeCache.challenge!.scope);
```

Comment:
It's hard to tell from this diff, but is this policy created for each new client, or shared? In .NET, our could be shared so we actually need to key our challenge cache (i.e. get specific ones, since there could be multiple). I don't see that possible here, so I'm hoping this is a new policy per client.

Reply:
This is new policy per client